src/analyses/basePriv.ml

```diff
@@ -0,0 +1,1664 @@
+    d
+
+  let write_global ask getg sideg (st: BaseComponents (D).t) x v =
+    let s = current_lockset ask in
+    let (vv, l) = st.priv in
+    let v' = L.fold (fun m _ acc ->
+        V.add m (MustVars.add x (V.find m acc)) acc
+      ) l vv
+    in
+    let cpa' = CPA.add x v st.cpa in
+    if not (!GU.earlyglobs && is_precious_glob x) then (
+      let v = distr_init getg x v in
+      sideg (mutex_global x) (G.create_weak (GWeak.singleton s v))
+    );
+    {st with cpa = cpa'; priv = (v', l)}
+
+  let lock ask getg (st: BaseComponents (D).t) m =
+    let s = current_lockset ask in
+    let (v, l) = st.priv in
+    let v' = V.add m (MustVars.empty ()) v in
+    let l' = L.add m (MinLocksets.singleton s) l in
+    {st with priv = (v', l')}
+
+  let unlock ask getg sideg (st: BaseComponents (D).t) m =
+    let s = Lockset.remove m (current_lockset ask) in
+    let is_in_G x _ = is_global ask x in
+    let side_cpa = CPA.filter is_in_G st.cpa in
+    let side_cpa = CPA.mapi (fun x v ->
+        let v = distr_init getg x v in
+        v
+      ) side_cpa
+    in
+    sideg (mutex_addr_to_varinfo m) (G.create_sync (GSync.singleton s side_cpa));
+    (* m stays in v, l *)
+    (* TODO: why is it so imprecise now? *)
+    st
+
+  let sync ask getg (st: BaseComponents (D).t) reason =
+    match reason with
+    | `Return -> (* required for thread return *)
+      begin match ThreadId.get_current ask with
+        | `Lifted x when CPA.mem x st.cpa ->
+          let v = CPA.find x st.cpa in
+          ({st with cpa = CPA.remove x st.cpa}, [(mutex_global x, G.create_weak (GWeak.singleton (Lockset.empty ()) v))])
+        | _ ->
+          (st, [])
+      end
+    | `Normal
+    | `Join (* TODO: no problem with branched thread creation here? *)
+    | `Init
+    | `Thread ->
+      (st, [])
+
+  let escape ask getg sideg (st: BaseComponents (D).t) escaped =
+    let cpa' = CPA.fold (fun x v acc ->
+        if EscapeDomain.EscapedVars.mem x escaped then (
+          sideg (mutex_global x) (G.create_weak (GWeak.singleton lockset_init v));
+          CPA.remove x acc
+        )
+        else
+          acc
+      ) st.cpa st.cpa
+    in
+    {st with cpa = cpa'}
+
+  let enter_multithreaded ask getg sideg (st: BaseComponents (D).t) =
+    CPA.fold (fun x v (st: BaseComponents (D).t) ->
+        if is_global ask x then (
+          sideg (mutex_global x) (G.create_weak (GWeak.singleton lockset_init v));
+          {st with cpa = CPA.remove x st.cpa}
+        )
+        else
+          st
+      ) st.cpa st
+
+  let threadenter = startstate_threadenter startstate
+end
+
+module WriteCenteredGBase =
+struct
+  open Locksets
+
+  module GWeakW = MapDomain.MapBot (Lockset) (VD)
+  module GSyncW = MapDomain.MapBot (Lockset) (CPA)
+
+  (* weak: G -> (S:2^M -> (W:2^M -> D)) *)
+  (* sync: M -> (S:2^M -> (W:2^M -> (G -> D))) *)
+  include AbstractLockCenteredGBase (GWeakW) (GSyncW)
+end
+
+module WriteCenteredD =
+struct
+  open Locksets
+
+  module W =
```

Comment:
Why does W and P have a different order here? The order should be the same, the only difference is the start value:
For `W` we start with `\bot` = `{}`, whereas for `P` we start with `\top` = `{{}}`.

But the order is always the same `the set of all upward-closed subsets of M, ordered by subset inclusion.`

Reply:
Or this exactly what we achieve here, i.e., non-present bindings are `bot` in W, and `top` in P?

Reply:
> the set of all upward-closed subsets of M, ordered by subset inclusion

This gives the order of `MinLocksets`.



> Why does `W` and `P` have a different order here?

The order for the map domain itself determines what happens when two branches are joined but one contains the key while the other doesn't. The `MapBot` vs `MapTop` determines what the default value for missing keys (and by extension behavior of the join with those missing keys) is. And hence the map domain orderings match up with the assumed start values of the inner `MinLocksets`.

Reply:
Alright, thanks for clarifying!